src/coreclr/jit/switchrecognition.cpp

```diff
@@ -132,6 +139,204 @@ bool IsConstantTestCondBlock(const BasicBlock* block,
+    {
+        JITDUMP("Couldn't extract the switch's variable - bail out.\n");
+        return false;
+    }
+
+    // TODO-JumpTable: Currently, we only expand the default target, but it should be possible
+    // to traverse all non-default edges as well (if they point to similar value tests).
+    // It shouldn't be too hard to implement (might be TP costly, though).
+
+    BBswtDesc* switchTargets = block->GetSwitchTargets();
+    if (!switchTargets->bbsHasDefault)
+    {
+        JITDUMP("Switch doesn't have a default target - bail out.\n");
+        return false;
+    }
+
+    // Now let's see if we can merge the value test in the edgeToExpand's destination block
+    // with the switch block.
+    FlowEdge*   edgeToExpand    = switchTargets->getDefault();
+    BasicBlock* edgeToExpandDst = edgeToExpand->getDestinationBlock();
+
+    // IsConstantTestCondBlock will tell us if the default target is a value test block (e.g. "x == cns")
+    // TODO-JumpTable: Add support for more complex tests like "x > 10 && x < 20"
+    // Also, we should be able to merge two BBJ_SWITCH on the same variable. Roslyn may create such shapes.
+    FlowEdge* testPasses;
+    FlowEdge* testFails;
+    bool      isReversed;
+    GenTree*  variableNode;
+    ssize_t   cns;
+    if (!IsConstantTestCondBlock(edgeToExpandDst, false, &testPasses, &testFails, &isReversed, &variableNode, &cns))
+    {
+        JITDUMP("The default target is not performing a value test - bail out.\n");
+        return false;
+    }
+
+    BasicBlock* testPassesBb = testPasses->getDestinationBlock();
+    BasicBlock* testFailsBb  = testFails->getDestinationBlock();
+
+    if (!GenTree::Compare(variableNode, switchOp))
+    {
+        JITDUMP("The default target is performing a value test on a different variable - bail out.\n");
+        return false;
+    }
+
+    // We're less conservative than Roslyn, but we still have some limits
+    if (static_cast<size_t>(cns + switchTargetOffset) > SWITCH_MAX_DISTANCE)
+    {
+        JITDUMP("Switch value is out of range - bail out.\n");
+        return false;
+    }
+
+    // We need to normalize the test value to the switch's range
+    const ssize_t newTestValue = cns + switchTargetOffset;
+
+    unsigned   oldJumpCnt = switchTargets->bbsCount;
+    unsigned   newJumpCnt = max(oldJumpCnt, (unsigned)(newTestValue + 1 + (switchTargets->bbsHasDefault ? 1 : 0)));
+    FlowEdge** oldJumpTab = switchTargets->bbsDstTab;
+    FlowEdge** newJumpTab = new (this, CMK_FlowEdge) FlowEdge*[newJumpCnt];
+
+    //
+    // Update likelihoods
+    //
+    weight_t edgeToExpandLikelihood = edgeToExpand->getLikelihood();
+    weight_t newTestLikelihood      = edgeToExpandLikelihood * testPasses->getLikelihood();
+    weight_t newDefLikelihood       = edgeToExpandLikelihood * testFails->getLikelihood();
+
+    // However, it is tricky if testPasses or testFails are already presented in the jump table.
```

Comment:
This might create a new dominant case for switch peeling (which happens later during block layout). If the default case was the most likely case before, and the new non-default case is now the most likely case and is reachable by just one case value, consider setting `bbsHasDominantCase` and `bbsDominantCase`.

Reply:
> This might create a new dominant case for switch peeling (which happens later during block layout)

I don't think this is true. We currently run switch recognition after `optOptimizeLayout`, and the late reordering pass in the backend only reorders blocks, without any of the additional opt passes like `fgOptimizeSwitchJumps`. Perhaps we should consider moving switch recognition before layout?

Reply:
> Perhaps we should consider moving switch recognition before layout?

Afair, the order of when to do it was driven by diffs, I can check again

Reply:
FWIW, I plan to move layout completely to the backend, and leave all the flow opts (like switch peeling) where they are now, so it might be easier to flip `optOptimizeLayout` and `optSwitchRecognition` after that change goes in.

Reply:
Thanks! I've re-ordered the phases, diffs didn't change much